### PR TITLE
SSEMR-238 Get actual EAC sessions dates and not the encounter dates

### PR DESCRIPTION
**Pull request by Michaelndula**
@moshonk please help review

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved SQL queries for evaluating last EAC dates, enhancing data accuracy.
  
- **New Features**
	- Enhanced data retrieval logic for last EAC dates by utilizing substring extraction methods.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java [hunk 1]
```diff
@@ -37,10 +37,8 @@ public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationC
 	        throws EvaluationException {
 		EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
 		
-		String qry = "SELECT client_id, DATE_FORMAT(max(encounter_datetime), '%Y-%m-%d')  as lastEac1Date FROM ssemr_etl.ssemr_flat_encounter_high_viral_load "
-		        + " WHERE  date(encounter_datetime) <= date(:endDate) "
-		        + " AND eac_session = 'First EAC Session' "
-		        + " GROUP BY client_id;";
+		String qry = "SELECT client_id, DATE_FORMAT(MID(MAX(CONCAT(encounter_datetime, adherence_date)), 20), '%Y-%m-%d') AS lastEac1Date "
```
Comment (moshonk) [inline api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java:40]: Why did you remove the session filter? Given that the form used for the various sessions is the same, how will you now differentiate between session1, session2, session3 etc?

File: api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC2DateDataEvaluator.java [hunk 2]
```diff
@@ -36,10 +36,8 @@ public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationC
 	        throws EvaluationException {
 		EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
 		
-		String qry = "SELECT client_id, DATE_FORMAT(max(encounter_datetime), '%Y-%m-%d') as lastEac2Date FROM ssemr_etl.ssemr_flat_encounter_high_viral_load "
-		        + " WHERE  date(encounter_datetime) <= date(:endDate) "
-		        + " AND eac_session = 'Second EAC Session' "
-		        + " GROUP BY client_id;";
+		String qry = "SELECT client_id, DATE_FORMAT(MID(MAX(CONCAT(encounter_datetime, second_eac_session_date)), 20), '%Y-%m-%d') AS lastEac2Date "
```
Comment (moshonk) [inline api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC2DateDataEvaluator.java:39]: Similar comment as above

File: api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC3DateDataEvaluator.java [hunk 3]
```diff
@@ -36,10 +36,8 @@ public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationC
 	        throws EvaluationException {
 		EvaluatedPersonData c = new EvaluatedPersonData(definition, context);
 		
-		String qry = "SELECT client_id, DATE_FORMAT(max(encounter_datetime), '%Y-%m-%d') as lastEac3Date FROM ssemr_etl.ssemr_flat_encounter_high_viral_load "
-		        + " WHERE  date(encounter_datetime) <= date(:endDate) "
-		        + " AND eac_session = 'Third EAC Session' "
-		        + " GROUP BY client_id;";
+		String qry = "SELECT client_id, DATE_FORMAT(MID(MAX(CONCAT(encounter_datetime, third_eac_session_date)), 20), '%Y-%m-%d') AS lastEac3Date "
```
Comment (moshonk) [inline api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC3DateDataEvaluator.java:39]: Similar comment as above

File: api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java (hunk 1 above)
Comment (Michaelndula) [reply api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java:40]: @moshonk we now have different sections for the EAC sessions, where each EAC session has a date field.
Screenshot from 2024-10-10 02-50-48

We removed the dropdowns.

File: api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java [hunk 4]
```diff
@@ -37,10 +37,8 @@ public EvaluatedPersonData evaluate(PersonDataDefinition definition, EvaluationC
...
+		        + "FROM ssemr_etl.ssemr_flat_encounter_high_viral_load WHERE date(encounter_datetime) <= date(:endDate) GROUP BY client_id";
```
[the remark this answers is not in the document]
Comment (moshonk) [reply api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java:41]: The goal is to get the max value that is a concatenation of the two fields. The encounter_datetime is not nullable

File: api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java (hunk 1 above)
Comment (moshonk) [reply api/src/main/java/org/openmrs/module/ssemrreports/reporting/library/data/evaluator/LastEAC1DateDataEvaluator.java:40]: Alright @Michaelndula. looks good then.

**Pull request merged**